### PR TITLE
Fixes #520 

### LINKED ISSUE #520: When navigating to `/<username>/sketches/<random_string>`, does not return 404 if the sketch does not exist
#### Nature of issue?

- [X] Found a bug

#### Details about the bug: 

- Web browser and version: 
- Operating System: MacOSX
- Steps to reproduce this:

Go to https://alpha.editor.p5js.org/cassie/sketches/random_string_that_does_not_match_an_id. The editor does not return a 404, but instead just serves the default sketch, one that would be returned if you had created a new sketch. The expected behavior is that the user would be directed to a 404 page, as it is unclear that the user has done anything wrong right now.

### PR DESCRIPTION
**Pull request by shinytang6**
Before your pull request is reviewed and merged, make sure you

* [x] there are no linting errors -- `npm run lint`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master. If you're asked to make more changes make sure you rebase onto master then too!
* [x] pull request is descriptively named and links to an issue number, i.e. `Fixes #123`

Thank you! Fixes #520

### CODE REVIEW

**Comment by shinytang6** [comment]
@catarak Can not think of a better way. please review:)

**Comment by catarak** [comment]
since all of the links on the front end are generated by the front end, i think that the 404 page should be returned from the server (i.e. in `server.routes.js`). also, this is maybe not totally clear, but you changed one of the API routes endpoints, which don't need to change! the API urls should return a 404 status code (but still json) and the website endpoints should redirect to the 404 page. sorry this was not clear in the issue!

**Comment by shinytang6** [comment]
get√ but l think maybe this api still needs the param `username`? It only checks `projectId` whether who owns it

**Comment by catarak** [comment]
the `projectId`s are unique for each sketch so the `username` shouldn't be necessary! though maybe i'm confused about what you mean.

**Comment by shinytang6** [comment]
emmm what l think is though the projectId is unique, the source doesn’t check whether it belongs a existing user

**Comment by catarak** [comment]
i see what you're saying! there are a few different routes here:
 
from `server.routes.js`:
`/projects/:project_id`
`/:username/sketches/:project_id`
from `project.routes.js`:
`/api/projects/:project_id`

in this PR you changed the last route to `/api/:username/projects/:project_id`, which doesn't need to be changed. as of right now, since the API is not public, front end users are never going to api routes, and they are generated by the front end. it is important that the first two routes redirect to 404 pages if the `:project_id` or `:username` does not exist. does that make sense?

**Comment by shinytang6** [comment]
cool :)

File: server/routes/server.routes.js [hunk 1]
```diff
@@ -20,11 +21,17 @@ router.get('/signup', (req, res) => {
...
 router.get('/projects/:project_id', (req, res) => {
-  res.send(renderIndex());
+  projectExists(req.params.project_id, exists => (
+    exists ? res.send(renderIndex()) : get404Sketch(html => res.send(html))
+  ));
 });
 
 router.get('/:username/sketches/:project_id', (req, res) => {
-  res.send(renderIndex());
+  userExists(req.params.username, exists => (
```
Comment (catarak) [inline server/routes/server.routes.js]: can you also check if the sketch belongs to the user here?

Comment (shinytang6) [reply server/routes/server.routes.js]: done :) Tell me if there are any other mistakes!

**Pull request merged**